### PR TITLE
Recorder refactoring that improves maintainability

### PR DESCRIPTION
**Pull request by rluders**
This PR implement improves the recorder maintainability, giving it the possibility to have multiple record drivers and by isolating code responsibility.

## Categories

- [ ] Bugfix
- [X] Enhancement
- [ ] Backporting
- [x] Others (CI, Infrastructure, Documentation)

## Sample archive

No.

## Documentation

No.

## Unit Tests

No.

## Privacy

Yes. There are no sensitive data in the newly collected information.

## Changelog

No.

## References

https://issues.redhat.com/browse/CCXDEV-2108

### CODE REVIEW

**Comment by tremes** [comment]
@rluders no need to create a new bugzilla to merge to master

**Comment by rluders** [comment]
@tremes all right, that is true. :man_facepalming: I always forget about it.

**Comment by 0sewa0** [comment]
tested it locally everything is working fine, changes make sense

/lgtm

File: pkg/recorder/recorder_test.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	}
+}
+
+func TestRecord(t *testing.T) {
+	rec := newRecorder()
+	err := rec.Record(record.Record{
+		Name: "config/mock1",
+		Item: tests.RawReport{Data: "mock1"},
+	})
+	assert.Nil(t, err)
```
Comment (tremes) [inline pkg/recorder/recorder_test.go:50]: This doesn't test much IMO. I think we can assert that the recorder actually has the new record.

File: pkg/recorder/recorder_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		Name:        "config/mock1",
+		Item:        tests.RawReport{Data: "mock1"},
+		Fingerprint: "abc",
+	})
+	err := rec.Record(record.Record{
+		Name:        "config/mock1",
+		Item:        tests.RawReport{Data: "mock1"},
+		Fingerprint: "abc",
+	})
+	assert.Nil(t, err)
```
Comment (tremes) [inline pkg/recorder/recorder_test.go:65]: Same here. We can assert that there's only one (or two?) record available.

File: pkg/recorder/recorder_test.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+func TestFlush(t *testing.T) {
+	rec := newRecorder()
+	for i := range []int{1, 2, 3} {
+		_ = rec.Record(record.Record{
+			Name: fmt.Sprintf("config/mock%d", i),
+			Item: tests.RawReport{Data: "mockdata"},
+		})
+	}
+	err := rec.Flush(context.TODO())
+	assert.Nil(t, err)
```
Comment (tremes) [inline pkg/recorder/recorder_test.go:86]: Same here. I guess the records should be cleared from recorder after flush call.

File: pkg/recorder/recorder_test.go (hunk 1 above)
Comment (rluders) [reply pkg/recorder/recorder_test.go:50]: So, actually, there isn't a way to get this info from the recorder. Unless if I change something to expose this information to the Recorder structure, or return the recorder item.

File: pkg/recorder/recorder_test.go (hunk 2 above)
Comment (rluders) [reply pkg/recorder/recorder_test.go:65]: The same thing above.

File: pkg/recorder/recorder_test.go (hunk 3 above)
Comment (rluders) [reply pkg/recorder/recorder_test.go:86]: The same comments above. And the recorder gets clear by [this line](https://github.com/openshift/insights-operator/pull/337/files/70480deff57f2d0eaf7b0483719b405b09411b92#diff-93cbf5b12a5398c622a4f899c81da8bdca98a02a5cf178e90f1b6b43e6b7005cR98).

File: pkg/recorder/recorder_test.go (hunk 1 above)
Comment (tremes) [reply pkg/recorder/recorder_test.go:50]: Well you can do something like `assert.Equal(t, 1, len(rec.records))` right

Comment (rluders) [reply pkg/recorder/recorder_test.go:50]: Oh... You are right. I totally forgot about the fact that 'cause the tests are in the same package we can access non-public struct props.

File: pkg/recorder/recorder_test.go (hunk 3 above)
Comment (rluders) [reply pkg/recorder/recorder_test.go:88]: Yep. Bug confirmed. Let me fix it.

**Comment by tremes** [comment]
Please suqash the commits :)

**Comment by tremes** [comment]
Let me experiment:
/tide/merge-method-merge

**Pull request merged**